stm32-modules/heater-shaker/firmware/freertos_comms_task.cpp

```diff
@@ -83,20 +87,50 @@ static std::array<StackType_t, stack_size> stack;
 static StaticTask_t
     data;  // NOLINT(cppcoreguidelines-avoid-non-const-global-variables)
 
+extern "C" {
+extern __ALIGN_BEGIN uint8_t
+    // NOLINTNEXTLINE(cppcoreguidelines-avoid-non-const-global-variables)
+    USBD_CDC_CfgHSDesc[USB_CDC_CONFIG_DESC_SIZ] __ALIGN_END;
+extern __ALIGN_BEGIN uint8_t
+    // NOLINTNEXTLINE(cppcoreguidelines-avoid-non-const-global-variables)
+    USBD_CDC_CfgFSDesc[USB_CDC_CONFIG_DESC_SIZ] __ALIGN_END;
+}
+
 // Actual function that runs in the task
 void run(void *param) {  // NOLINT(misc-unused-parameters)
-    static constexpr uint32_t delay_ticks = 1;
     auto *task_pair = static_cast<decltype(_tasks) *>(param);
     auto *local_task = task_pair->second;
+    auto *top_task = task_pair->first;
+    // This clears the capability bit that would be other sent upstream
+    // indicating we handle flow control line setting from host, which we don't,
+    // which leads to delays and annoying kernel messages. See
+    // stm32f303-bsp/Middlewares/ST/STM32_USB_Device_Library/Class/CDC/Src/usbd_cdc.c:174
+    // for annotated descriptor definitions
+    // NOLINTNEXTLINE(cppcoreguidelines-avoid-magic-numbers)
+    USBD_CDC_CfgHSDesc[30] = 0;
+    // NOLINTNEXTLINE(cppcoreguidelines-avoid-magic-numbers)
+    USBD_CDC_CfgFSDesc[30] = 0;
     USBD_Init(&local_task->usb_handle, &CDC_Desc, 0);
     USBD_RegisterClass(&local_task->usb_handle, USBD_CDC_CLASS);
     USBD_CDC_RegisterInterface(&local_task->usb_handle,
                                &local_task->cdc_class_fops);
     USBD_SetClassConfig(&local_task->usb_handle, 0);
     USBD_Start(&local_task->usb_handle);
+    local_task->committed_rx_buf_ptr = local_task->rx_buf.committed()->data();
     while (true) {
-        USBD_CDC_ReceivePacket(&local_task->usb_handle);
-        vTaskDelay(delay_ticks);
+        auto length_to_xmit =
+            top_task->run_once(local_task->tx_buf.accessible()->data(),
+                               local_task->tx_buf.accessible()->size());
+        if (length_to_xmit != 0) {
+            local_task->tx_buf.swap();
+            USBD_CDC_SetTxBuffer(
```

Comment:
I'm having trouble finding this function and `USBD_CDC_TrasmitPacket`

Reply:
they are in the usb driver: https://github.com/STMicroelectronics/STM32CubeF3/blob/v1.11.2/Middlewares/ST/STM32_USB_Device_Library/Core/Inc/usbd_core.h

in general, functions that kind of look like C functions (they have their "namespace" baked into their function call) are going to be from st's driver package